### ISSUE TITLE
motor object .set_current_position leaves motor in 'set' state

### ISSUE THREAD

**Issue by dlzhu** [pcdshub/Bug-Reports-and-Requests#47]
Currently for some motors (XPP and XCS for sure, no idea about other hutches), in particular dumb motors, the .set_current_position function hangs in the terminal. When Ctr+C out of it, it leaves the motor in the 'set' state.

**Comment by ZryletTC** [comment]
That's very weird. From what I remember, that method is only three lines long. It'll be interesting to find out which one is stalling.

**Comment by dlzhu** [comment]
exactly, it's 3 caputs.

On Oct 6, 2020, at 11:36 AM, Tyler Pennebaker <notifications@github.com> wrote:

﻿

That's very weird. From what I remember, that method is only three lines long. It'll be interesting to find out which one is stalling.

**Comment by klauer** [comment]
Needs a `try: / finally:` block, apparently?

**Comment by ZryletTC** [comment]
@klauer Yeah that's what I was thinking but it'd also be nice to figure out why it's failing in the first place

**Comment by ZLLentz** [comment]
I've never trusted this method because it's always looked like it could cause trouble

**Comment by dlzhu** [comment]
Ping on this again :P can we allocate some more priority to get this fixed? Or make up a simpler, more bullet proof '3-line caput' script if we couldn't figure out what's wrong with the current one? This is a function that we use a lot, and currently all user motors are being added to the new hutch python ONLY, so there's not a 'workaround' :P

**Comment by mseaberg** [comment]
I found that putting a wait_for_connection after each of the first two puts, and taking out the wait=True in the put calls, seems to work well.

**Comment by ZLLentz** [comment]
linking this issue with https://github.com/pcdshub/pcdsdevices/issues/684

**Comment by ZLLentz** [comment]
So there are two problems here:
- `set_position` shouldn't use the set/use switch. It doesn't need to and it makes us susceptible to errors. This should be an easy fix.
- `EpicsSignal` puts are hanging. This is not easy to debug, but Matt's `wait_for_connection` fix suggests a few thing to look at.

**Comment by klauer** [comment]
Would your suggested alternative be to write directly to `.OFF`, @ZLLentz? If so, there are some fundamental differences using it versus the current `.SET`, and we should be certain that we are OK with that for all of our motor record-based motors.

**Comment by ZLLentz** [comment]
@klauer does using the `.SET` switch cause some hardware-specific behavior to happen?

**Comment by klauer** [comment]
[Yes](https://epics.anl.gov/bcda/synApps/motor/motorRecord.html):
> When SET = 1 ("Set"), writes to the dial-coordinate drive field (DVAL) and to the raw drive field (RVAL) cause a new raw motor position to be **loaded into the hardware** without any change to the user-coordinate drive field (VAL).

Writing to `.OFF` does not do this - which _can_ be OK, but I don't think is a safe assumption for a shared library such as ophyd to make.

**Comment by ZLLentz** [comment]
REF: the previous code that Matt reported working, from the lv1818 experiment file
```
    def set_current_position(self, motor, value):
        motor.set_use_switch.put(1)
        motor.set_use_switch.wait_for_connection()
        motor.user_setpoint.put(value, force=True)
        motor.user_setpoint.wait_for_connection()
        motor.set_use_switch.put(0)
```

**Comment by ZryletTC** [comment]
I just realized today that we never did anything with this. I've temporarily implemented this in a dev version of ophyd for xcs, but this should be raised upstream, right?

**Comment by ZLLentz** [comment]
Yes, I didn't realize we hadn't completed this
